x/airdrop/keeper/keeper.go

```diff
@@ -0,0 +1,200 @@
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	"github.com/cosmos/cosmos-sdk/x/airdrop/types"
+	authtypes "github.com/cosmos/cosmos-sdk/x/auth/types"
+	paramtypes "github.com/cosmos/cosmos-sdk/x/params/types"
+)
+
+type Keeper struct {
+	bankKeeper       types.BankKeeper
+	cdc              codec.BinaryMarshaler
+	storeKey         sdk.StoreKey
+	feeCollectorName string
+	paramSpace       paramtypes.Subspace
+}
+
+type FundPair struct {
+	Fund    types.Fund
+	Account sdk.AccAddress
+}
+
+func NewKeeper(cdc codec.BinaryMarshaler, storeKey sdk.StoreKey, paramSpace paramtypes.Subspace, bankKeeper types.BankKeeper, feeCollectorName string) *Keeper {
+
+	// set KeyTable if it has not already been set
+	if !paramSpace.HasKeyTable() {
+		paramSpace = paramSpace.WithKeyTable(types.ParamKeyTable())
+	}
+
+	return &Keeper{
+		bankKeeper:       bankKeeper,
+		cdc:              cdc,
+		storeKey:         storeKey,
+		feeCollectorName: feeCollectorName,
+		paramSpace:       paramSpace,
+	}
+}
+
+func (k Keeper) AddFund(ctx sdk.Context, sender sdk.AccAddress, fund types.Fund) error {
+	params := k.GetParams(ctx)
+	if !params.IsAllowedSender(sender) {
+		return sdkerrors.Wrapf(
+			sdkerrors.ErrConflict,
+			"Non-whitelist sender %s", sender.String(),
+		)
+	}
+
+	// validate that the fund we are adding is correct
+	err := fund.ValidateBasic()
+	if err != nil {
+		return err
+	}
+
+	err = k.bankKeeper.SendCoinsFromAccountToModule(ctx, sender, types.ModuleName, sdk.NewCoins(*fund.Amount))
+	if err != nil {
+		return err
+	}
+
+	return k.setFund(ctx, sender, fund, false);
+}
+
+func (k Keeper) UpdateFund(ctx sdk.Context, sender sdk.AccAddress, fund types.Fund) error {
+	return k.setFund(ctx, sender, fund, true);
+}
+
+func (k Keeper) setFund(ctx sdk.Context, sender sdk.AccAddress, fund types.Fund, shouldExist bool) error {
+	key := types.GetActiveFundKey(sender)
+
+	// check that we do not already have fund from this account
+	store := ctx.KVStore(k.storeKey)
+
+	// check to see if the fund should exist or not
+	if shouldExist {
+		if !store.Has(key) {
+			return sdkerrors.Wrapf(sdkerrors.ErrConflict, "Fund from sender already exists")
+		}
+
+		// if a fund is updated with a zero or negative remaining amount then simple delete the entry
+		if fund.Amount.IsNegative() || fund.Amount.IsZero() {
+			store.Delete(key) // remove the entry
+		} else {
+			store.Set(key, k.cdc.MustMarshalBinaryBare(&fund)) // update the entry
+		}
+
+	} else {
+		if store.Has(key) {
+			return sdkerrors.Wrapf(sdkerrors.ErrConflict, "Fund from sender already exists")
+		}
+
+		// update the fund
+		store.Set(key, k.cdc.MustMarshalBinaryBare(&fund))
+	}
+
+	return nil
+}
+
+func (k Keeper) GetFund(ctx sdk.Context, sender sdk.AccAddress) (*types.Fund, error) {
+	key := types.GetActiveFundKey(sender)
+
+	store := ctx.KVStore(k.storeKey)
+	bz := store.Get(key)
+	if bz == nil {
+		return nil, nil
+	}
+
+	fund := &types.Fund{}
+	k.cdc.MustUnmarshalBinaryBare(bz, fund)
```

Comment:
Actually seems to be a convention of the cosmos-sdk to prefer this panic route for serialisation issues. Not really sure was just following the convention to be honest. I can change.

Reply:
well, there are 2 cases really:
- if this code is reached only by CLI kind of calls, that's fine to panic, won't hurt anything
- if this is reached by some server/service, then it's bad as it will kill the thing unless there's some panic recovery handler, but that's unlikely.

in general, it's best to avoid panic as much as possible given we can't really predict where this will be used, when possible, bubble errors all the way up, and let the client panic if he'd like to. 

Now since the keeper marshaling and unmarshalling is all internal, there are low chances it can't unmarshal what it marshaled earlier, so no big deal to leave it like that, it just looks more fragile without real impact I think.